### PR TITLE
[mlir][loops] Add getters for multi dim loop variables in `LoopLikeOpInterface`

### PR DESCRIPTION
**Pull request by srcarroll**
This patch adds `getLoopInductionVars`, `getLoopLowerBounds`, `getLoopBounds`, `getLoopSteps` interface methods to `LoopLIkeOpInterface`.  The corresponding single value versions have been moved to shared class declaration and have been implemented based on the new interface methods.

### CODE REVIEW

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp
Comment (srcarroll) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:2464]: i have a feeling this won't fly, so i'm happy to hear thoughts

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (Hardcode84) [inline mlir/include/mlir/Interfaces/LoopLikeInterface.td:237]: this code will create a temporary `SmallVector` just to get size and then another one on the next line, better to cache.

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (Hardcode84) [inline mlir/include/mlir/Interfaces/LoopLikeInterface.td:131]: This is more of a question to the audience: there are 2 possible cases when empty vector is returned:
1. Cannot determine loop bounds (`scf.while`)
2. Some degenerate case, when loop has 0 dimensions.

Probably makes sense to distinguish those 2 cases and return `optional<SmallVector<OpFoldResult>>`.

Comment (srcarroll) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td:131]: sounds reasonable to me. should i wait to hear from others on this before i change?

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Interfaces/LoopLikeInterface.td:98]: Probable not a good idea to return a `ValueRange` unless you are sure this is always contiguous.

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (MaheshRavishankar) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td:131]: `optional<SmallVector<OpFoldResult>>` makes sense to me. Should be true for all of the methods added here.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:2464]: Can you explain a little bit as to why the semantics are not quite right?

**Review by MaheshRavishankar — COMMENTED** [review]
cc @jtuyls  .

I am fine with this approach modulo inputs from others. @jtuyls had a separate approach where he was thinking of defining a new interface that would create an inherited interface for loops with induction variables. I am partial to the approach in this PR, but the other approach is valid as well.

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (srcarroll) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td:98]: i actually started with `SmallVector<Value>` but was running into an error saying something about not being able to expand the vector. i didn't really understand it to be honest. i'll try again

Comment (MaheshRavishankar) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td:98]: `ValueRange` is essentially a reference. Returning a reference is safe only if you know it is part of the `operands` or `BlockArguments` of the operation.

**Comment by srcarroll** [comment]
> Can you explain a little bit as to why the semantics are not quite right?
Its supposed  to return mixed values, but here i'm only returning a non empty vector if the result is a constant. i did this to keep the same semantics for the older interface methods on the `affine::ForOp`.  but really it should return any OpFoldResult. atm this doesn't affect anything because this op doesn't use the new methods anyway

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp
Comment (srcarroll) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:2464]: Its supposed to return mixed values, but here i'm only returning a non empty vector if the result is a constant. i did this to keep the same semantics for the older interface methods on the affine::ForOp. but really it should return any OpFoldResult. atm this doesn't affect anything because this op doesn't use the new methods anyway

Comment (srcarroll) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:2464]: my issue was that i didn't really fully understand how affine loop bounds work compared to scf. because of the whole `AffineBound` thing, which i'm unfamiliar with

**Comment by srcarroll** [comment]
> optional<SmallVector<OpFoldResult>> makes sense to me. Should be true for all of the methods added here.

so i'm definitely in agreement with both of you on this. but this requires a lot of changes because of the way i have named the functions. For example `scf::ForallOp` already had `getMixedUpperBound`, etc. I kept this naming convention for the interface methods.  The issue now is that everywhere this function was called, now I have to check if there is a value and then get the value. I'm personally fine with this, but wonder if anyone will be annoyed by the inconvenience. 

Any thoughts?

**Comment by srcarroll** [comment]
> would it be ok to just have another function for this op with the same name, but just return SmallVector<Value> and reuse the interface version

nevermind. realized you can't do that

**Comment by srcarroll** [comment]
i'm thinking i should keep the `getMixed` names for `scf::ForallOp` and change the interface names

**Comment by srcarroll** [comment]
> I am fine with this approach modulo inputs from others. @jtuyls had a separate approach where he was thinking of defining a new interface that would create an inherited interface for loops with induction variables. I am partial to the approach in this PR, but the other approach is valid as well.

i took a brief look at this. maybe i'm missing something, but i'm not much of a fan for a whole new interface just to get the multi induction/bounds/etc.

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (srcarroll) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td]: resolved the errors. but to be honest, i dont quite know why what i did fixed it. could someone take a look at this commit to make sure i didn't do something stupid. https://github.com/llvm/llvm-project/pull/94516/commits/1babe681d7858a4992303c62e22684cb73d82472

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:2467]: I dont know if you need to worry about it? the induction variable is just an SSA value.

Comment (srcarroll) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:2467]: i'm refering to the bounds and steps

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (makslevental) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td]: i left comments on that commit but from experience I know that you'll have a really hard time finding them (they will not appear here) so https://github.com/llvm/llvm-project/commit/1babe681d7858a4992303c62e22684cb73d82472

**Comment by jtuyls** [comment]
> > I am fine with this approach modulo inputs from others. @jtuyls had a separate approach where he was thinking of defining a new interface that would create an inherited interface for loops with induction variables. I am partial to the approach in this PR, but the other approach is valid as well.
> 
> i took a brief look at this. maybe i'm missing something, but i'm not much of a fan for a whole new interface just to get the multi induction/bounds/etc. but i'm certainly no authority on this

Initially, I used `optional<>`/empty list/setters returning a `LogicalResult` on all methods related to induction variables, but imo, having to do that is demonstrating that not all loops are equal (or all are, but some more equal than others) and that there is a distinction between loops with induction variables and ones without. By creating a new interface  you don't need the surrounding logic everywhere these methods are used to evaluate the return value and check whether it was valid or not to use this method on that operation or not. Also, this let's you reuse the existing induction variable related methods inside `scf.Forall`. But mainly, I think the point of an interface is to have a contract on what functionality is available for operations implementing the interface. Using `optional<>` on a set of methods, not a one-off, defeats that purpose and breaks the contract imo.

File: mlir/include/mlir/Dialect/SCF/IR/SCFOps.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/SCF/IR/SCFOps.td:516]: I am not sure `assert` is a good idea. This some time leads to spurious failures. What is reason where it wont have a value.

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Interfaces/LoopLikeInterface.td:236]: We should just deprecate these methods. THey never made sense to me.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Just marking requesting changes since I see an `assert`.

File: mlir/include/mlir/Dialect/Affine/IR/AffineOps.td
Comment (srcarroll) [reply mlir/include/mlir/Dialect/Affine/IR/AffineOps.td:674]: well i did this because i called the interface method `getSteps` which would conflict with the generated getter for this operand. probably a bad idea but not sure what else to do (couldn't come up with better naames)

File: mlir/include/mlir/Dialect/SCF/IR/SCFOps.td
Comment (srcarroll) [reply mlir/include/mlir/Dialect/SCF/IR/SCFOps.td:516]: there should be no reason, which is why i made it an assert. just a bug catcher in case other things change

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (srcarroll) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td:236]: i agree. but i think that would require a whole lot of changes that i'm not keen on doing right now

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (srcarroll) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td]: thanks!

File: mlir/include/mlir/Dialect/Affine/IR/AffineOps.td
Comment (MaheshRavishankar) [reply mlir/include/mlir/Dialect/Affine/IR/AffineOps.td:674]: Maybe make it `getLoopSteps` in the interface?

Comment (srcarroll) [reply mlir/include/mlir/Dialect/Affine/IR/AffineOps.td:674]: works for me. thanks :)

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:2467]: Yeah, I dont know either. @bondhugula, @Groverkss or @stephenneuendorffer   can you guys comment?

**Review by MaheshRavishankar — COMMENTED** [review]
Ok, I can live with the assert maybe, but these asserts have caused a lot of heart-ache downstream. I am fine with this. Giving a chance for other folks to weigh in. Maybe @matthias-springer or @ftynse

**Comment by Hardcode84** [comment]
> Initially, I used `optional<>`/empty list/setters returning a `LogicalResult` on all methods related to induction variables, but imo, having to do that is demonstrating that not all loops are equal (or all are, but some more equal than others) and that there is a distinction between loops with induction variables and ones without. By creating a new interface you don't need the surrounding logic everywhere these methods are used to evaluate the return value and check whether it was valid or not to use this method on that operation or not. Also, this let's you reuse the existing induction variable related methods inside `scf.Forall`. But mainly, I think the point of an interface is to have a contract on what functionality is available for operations implementing the interface. Using `optional<>` on a set of methods, not a one-off, defeats that purpose and breaks the contract imo.

Thanks for the explanation, this actually now makes sense to me to separate interfaces to cover the case when we cannot get the bounds. Both approaches can work, I won't block this PR is everyone else is fine with `optional<>`.

**Comment by srcarroll** [comment]
> Thanks for the explanation, this actually now makes sense to me to separate interfaces to cover the case when we cannot get the bounds. Both approaches can work, I won't block this PR is everyone else is fine with optional<>.

I'm still impartial to it, but i'm certainly no authority. would like to hear what others think too. 

I could maybe see the new interface being a kind of specialization of the existing `LoopLikeOpInterface`. but i really see no advantage of completely separate interface.

**Comment by srcarroll** [comment]
> I could maybe see the new interface being a kind of specialization of the existing LoopLikeOpInterface

I see this was already suggested https://github.com/llvm/llvm-project/pull/93781#discussion_r1621016980

**Comment by srcarroll** [comment]
>  but these asserts have caused a lot of heart-ache downstream

yah i've definitely been burned by people abusing asserts (using them over proper error checks), so i understand where you're coming from. but hopefully it's agreed that's not the case here. again it's purely for bug catching and not intended as an error check because the assert should never fail unless someone else modifies the code. in which case the assert would help debug

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (ftynse) [inline mlir/include/mlir/Interfaces/LoopLikeInterface.td:123]: Nit: why remove the leading `::` before `std` here but not elsewhere?

File: mlir/unittests/Dialect/SCF/LoopLikeSCFOpsTest.cpp
Comment (ftynse) [inline mlir/unittests/Dialect/SCF/LoopLikeSCFOpsTest.cpp:45]: The check above should become an `ASSERT_TRUE` so it exits the current function. Otherwise we will get a real assertion here. Same below.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp
Comment (ftynse) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:2467]: The bound is computed as an (affine) function of multiple values, which are passed as operands to the loop operation. It cannot be represented with one value since the result of that function is not materialized. All existing interface methods are "best effort", i.e., they return non-nullopt value if it makes sense for the specific loop. And should be documented as such if they aren't already. So it is fine to return nullopt for affine loops with non-constant upper bound.

If we _really_ needed that, we could add a `materializeLower/UpperBound(OpBuilder &)` method, but I doubt it's useful.

File: mlir/lib/Dialect/SCF/IR/SCF.cpp
Comment (ftynse) [inline mlir/lib/Dialect/SCF/IR/SCF.cpp:382]: I suspect using different number of stack-based elements in small vector between this and the return type will prevent move semantics here.

**Review by ftynse — APPROVED** [review]
This LGTM modulo nits.

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (ftynse) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td]: No idea what was happening there but 18446744073709551609 is `static_cast<uint64_t>(-6)` if that helps, and the default number of stack elements in a small vector for 64-bit element width is also 6. So I suspect the sign got inverted somewhere or the distance was taken in the wrong direction.

**Comment by ftynse** [comment]
> yah i've definitely been burned by people abusing asserts (using them over proper error checks), so i understand where you're coming from. but hopefully it's agreed that's not the case here. again it's purely for bug catching and not intended as an error check because the assert should never fail unless someone else modifies the code. in which case the assert would help debug

The rule of thumb is: if this can happen on IR that passed the verifier, it should be a diagnostic, otherwise it should be an assertion. Including the case when a downstream didn't implement the interface contract correctly. Asserts are for developers, including downstream developers. Diagnostics / null returns are for compiler users. This PR is using them correctly.

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (matthias-springer) [inline mlir/include/mlir/Interfaces/LoopLikeInterface.td:96]: Mention that:
* `std::nullopt` means that the loop does not expose loop induction variables via this interface. (E.g., because the loop op has no concept of "induction variables".)
* An empty vector means that the loop has no loop induction variables.

Same for the other getter methods.

File: mlir/include/mlir/Dialect/SCF/IR/SCFOps.td
Comment (matthias-springer) [inline mlir/include/mlir/Dialect/SCF/IR/SCFOps.td:855]: double `;`

**Review by matthias-springer — APPROVED** [review]

File: mlir/include/mlir/Dialect/SCF/IR/SCFOps.td
Comment (srcarroll) [reply mlir/include/mlir/Dialect/SCF/IR/SCFOps.td:855]: oh weird. i didn't know that even compiles

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (srcarroll) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td:123]: just a mistake. will fix

File: mlir/unittests/Dialect/SCF/LoopLikeSCFOpsTest.cpp
Comment (srcarroll) [reply mlir/unittests/Dialect/SCF/LoopLikeSCFOpsTest.cpp:45]: ah i didn't think about that. thanks

File: mlir/include/mlir/Interfaces/LoopLikeInterface.td
Comment (srcarroll) [reply mlir/include/mlir/Interfaces/LoopLikeInterface.td]: thanks for the explanation

**Comment by srcarroll** [comment]
thanks for all the reviews everyone. i think i addressed everyone's comments. hope i didn't miss anything

**Pull request merged**
